Introduce EXTRA_CONF_FILE and EXTRA_DTC_OVERLAY_CONFIG

tejlmand opened the pull request:
Fixes: #57775 

This PR introduces EXTRA_CONF_FILE  and EXTRA_DTC_OVERLAY_CONFIG while still keeping support for OVERLAY_CONFIG.

EXTRA_ZEPHYR_MODULES is introduced for consistency. Support for ZEPHYR_EXTRA_MODULES is kept.

It also updates the documentation to reflect those changes.

Note, this PR doesn't update snippets code or docs, because there is dedicated work ongoing for this, see #57661 , #57139, #57142

tejlmand committed:
cmake: extend zephyr_get() to support fetching of multiple variables

This commit extends zephyr_get() to support multiple variable names to
be fetched into a single variable.

Example:
  zephyr_get(FOO VAR FOO_A FOO_B FOO_C)

will lookup each FOO_A, FOO_B, FOO_C for supported scopes and return
the value in FOO of the first scope encountered having one of the
variables defined.

If MERGE is specified, then all scopes for all vars are looped and the
variable values are merged into FOO and returned.

This functionality will allow to deprecate user-facing settings while
ensuring that both the new and deprecated variables are considered
in zephyr_get() and also taking into consideration the scope with
highest precedence.

This allows Zephyr CMake to do:
  zephyr_get(FOO VAR FOO DEPRECATED_FOO)

  zephyr_get(BAR MERGE VAR BAR DEPRECATED_BAR)

This allows support of old and new setting for a given number of
releases.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: introduce EXTRA_CONF_FILE and deprecate OVERLAY_CONFIG

This commit introduces EXTRA_CONF_FILE and deprecates OVERLAY_CONFIG.

It has often caused confusion that OVERLAY_CONFIG adds extra
configuration fragments to the value of CONF_FILE (default: prj.conf),
but the similar named variable DTC_OVERLAY_FILE replaces the default
dtc overlay file used by the build system.

To remove such confusion, this commit introduces the EXTRA_ prefix in
front of CONF_FILE to clearly indicate it's purpose.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: introduce EXTRA_DTC_OVERLAY_FILE for devicetree

This commit introduces EXTRA_DTC_OVERLAY_FILE.

This allows users to specify additional devicetree overlays in addition
to a sample / board specific default devicetree overlay files.

It also allows snippets to provide extra devicetree overlays on addition
to sample / board specific overlays.

The name EXTRA_DTC_OVERLAY_FILE clearly indicates the purpose and
follows the naming scheme of EXTRA_ in front of variable name.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: update package helper description to use EXTRA_CONF_FILE var

Update package_helper.cmake to use EXTRA_CONF_FILE instead of deprecated
OVERLAY_CONFIG variable.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
sysbuild: update sysbuild to use EXTRA_CONF_FILE variable

Align sysbuild with Zephyr CMake so that sysbuild supports
SB_EXTRA_CONF_FILE. Keep support for the old SB_OVERLAY_CONFIG.

This ensures that sysbuild is consistent with Zephyr CMake configuration
file handling.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
samples: update nrf53_sync_rtc sample to EXTRA_CONF_FILE

Update nrf53_sync_rtc to use EXTRA_CONF_FILE instead of the old
OVERLAY_CONFIG.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
doc: introduce EXTRA_CONF_FILE and EXTRA_DTC_OVERLAY_FILE

Introduce EXTRA_CONF_FILE and EXTRA_DTC_OVERLAY_FILE and remove old
references to OVERLAY_CONFIG.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: align Zephyr module variable to EXTRA_ZEPHYR_MODULES

Align Zephyr modules with other user facing variables where settings
can be defined or extended, meaning Zephyr modules now supports:
ZEPHYR_MODULES and EXTRA_ZEPHYR_MODULES.

Support for ZEPHYR_EXTRA_MODULES is kept foir backward compatibility.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
doc: update Zephyr modules to use EXTRA_ZEPHYR_MODULES

Update Zephyr modules documentation to use EXTRA_ZEPHYR_MODULES.
This gives a more consistent naming scheme in Zephyr CMake build system.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

nordicjm commented:
> EXTRA_ZEPHYR_MODULES is introduced for consistency. Support for EXTRA_ZEPHYR_MODULES is kept.

Is this part correct?

nordicjm commented:
I count 147 uses of OVERLAY_CONFIG in `samples`, 37 in `tests` and 6 in `scripts` which will need updating

tejlmand commented on cmake/package_helper.cmake:
```diff
@@ -29,18 +29,18 @@
 # If you invoke CMake for 'hello_world' as:
 #
-#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DOVERLAY_CONFIG=foo.overlay
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DEXTRA_CONF_FILE=foo.overlay
 #
 # you just add the same argument to the helper like:
-#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DOVERLAY_CONFIG=foo.overlay \
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DEXTRA_CONF_FILE=foo.overlay \
 #           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_helper.cmake
```
very good point.

I was only updating the variable name, didn't pay attention to the file defined.
Will replace the old and misleading extension name.

tejlmand commented:
> I count 147 uses of OVERLAY_CONFIG in `samples`, 37 in `tests` and 6 in `scripts` which will need updating

Correct, but those are within sample.yaml / testcase.yaml and used by `twister`.
Because we keep support for the OVERLAY_CONFIG, then I prefer to wait with updating twister related files until we have the actual changes in place.

That way we ensure that try support for OVERLAY_CONFIG still works as expected, because twister still uses this, and then a followup PR adjusting sample.yaml / testcase.yaml to new variable name ensures that new name also works in all cases.

Finally it keeps number of files touched in this PR a bit smaller.

Note: some rst files are using OVERLAY_CONFIG, they should be updated in this PR, and I will do so.

tejlmand committed:
doc: samples: update OVERLAY_CONFIG to EXTRA_CONF_FILE

Update rst files to use EXTRA_CONF_FILE instead of OVERLAY_CONFIG.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> Note: some rst files are using OVERLAY_CONFIG, they should be updated in this PR, and I will do so.

done

mbolivar-nordic commented on cmake/modules/extensions.cmake:
```diff
@@ -2467,11 +2467,42 @@ function(zephyr_list transform list_var action)
+    if(DEFINED ${scope}_${variable})
+      set(${out} "${${scope}_${variable}}" PARENT_SCOPE)
+    else()
+      set(${out} PARENT_SCOPE)
```
non-blocking: `unset(${out} PARENT_SCOPE)` would be more readable IMO, but I see that `set(foo)` is used elsewhere too.

tejlmand replied:
In this particular case, where it is a function and the function sets the value of the return variable referenced by `out`.
Then I personally think it is cleaner to indicate that it sets the value of the return var to nothing (hence clears it), than to unset a return variable.

But I agree, it can be seen both ways.

mbolivar-nordic commented on cmake/modules/extensions.cmake:
```diff
@@ -2467,11 +2467,42 @@ function(zephyr_list transform list_var action)
 # Return the value of <variable> as local scoped variable of same name. If MERGE
 # is supplied, will return a list of found items.
 #
+# VAR can be used when the value of multiple variables must be considered, for
+# example to both consider FOO_A and FOO_B, use:
```
nonblocking: maybe this is a little better?

```
VAR can be used either to store the result in a variable with
a different name, or to look for values from multiple variables.
```

I could see this also being useful to do something like `zephyr_get(my_foo VAR FOO)`. It also makes it clearer that `<variable>` is not coupled to the `VAR` argument(s).

tejlmand replied:
thanks for the proposal

mbolivar-nordic commented on cmake/modules/extensions.cmake:
```diff
@@ -2524,59 +2558,41 @@ function(zephyr_get variable)
        ("${GET_VAR_SYSBUILD}" STREQUAL "GLOBAL" OR sysbuild_main_app)
     )
       get_property(sysbuild_${variable} TARGET sysbuild_cache PROPERTY ${variable})
```
Shouldn't this (and the previous `${variable}` on line 2556) refer to `GET_VAR_VAR` instead now?

tejlmand replied:
it should.

Moved inside the `foreach()` and variable --> var

mbolivar-nordic commented on cmake/modules/extensions.cmake:
```diff
@@ -2524,59 +2558,41 @@ function(zephyr_get variable)
+    foreach(var ${GET_VAR_VAR})
+      zephyr_var_name("${var}" "${scope}" expansion_var)
+      if(DEFINED expansion_var)
+        string(CONFIGURE "${expansion_var}" scope_value)
```
I'm curious about the use of CONFIGURE here. This needs to resolve `$CACHE` and `$ENV` references too, but I don't see any documentation in https://cmake.org/cmake/help/latest/command/configure_file.html that says that should work. Only `${VAR}` and `@VAR@` are explicitly supported. Is this relying on an implementation detail of cmake?

tejlmand replied:
That's a very good point.

So I opened this: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/8500

mbolivar-nordic commented on cmake/modules/extensions.cmake:
```diff
@@ -2524,59 +2558,41 @@ function(zephyr_get variable)
+            # Variable exists as current scoped variable, defined in a CMakeLists.txt
+            # file, however it is also set in environment.
+            # This might be a surprise to the user, so warn about it.
+            message(WARNING "environment variable '${variable}' is hiding local "
```
Shouldn't this be `${var}`, not `${variable}`? It seems so, since the `VAR` syntax decouples the `variable` name from the `VAR a b c [...]` names.

tejlmand replied:
nice catch.

Fixed.

mbolivar-nordic commented on cmake/modules/configuration_files.cmake:
```diff
@@ -116,4 +116,4 @@ DTC_OVERLAY_FILE=\"dts1.overlay dts2.overlay\"")
 zephyr_boilerplate_watch(DTC_OVERLAY_FILE)
 
-zephyr_get(OVERLAY_CONFIG SYSBUILD LOCAL)
+zephyr_get(EXTRA_CONF_FILE SYSBUILD LOCAL VAR EXTRA_CONF_FILE OVERLAY_CONFIG)
```
This seems to be a "silent" deprecation; will there be a way to warn users about this at some point? (Maybe it's later in the PR, but the commit message says that this commit is doing the deprecation of OVERLAY_CONFIG.)

mbolivar-nordic commented on cmake/modules/dts.cmake:
```diff
@@ -158,6 +158,15 @@ if(DTC_OVERLAY_FILE)
     )
 endif()
 
+if(EXTRA_DTC_OVERLAY_FILE)
```
This needs to be added to the optional variables list in the file's documentation comment above.

tejlmand replied:
together with the missing `DTC_OVERLAY_FILE` I assume 😉

tejlmand replied:
done

mbolivar-nordic commented on doc/build/dts/howtos.rst:
```diff
@@ -246,9 +246,13 @@ use as devicetree overlays:
 
+Extra devicetree overlays may be provided using ``EXTRA_DTC_OVERLAY_FILE`` which
+will still allow the build system to automatically use devicetree overlays
+described in the above steps.
```
please add:

```
The build system appends overlays specified in ``EXTRA_DTC_OVERLAY_FILE``
to the overlays in ``DTC_OVERLAY_FILE`` when processing devicetree overlays.
This means that changes made via ``EXTRA_DTC_OVERLAY_FILE`` have higher
precedence than those made via ``DTC_OVERLAY_FILE``.
```

(Order matters for the reasons described above!)

mbolivar-nordic commented on doc/develop/modules.rst:
```diff
@@ -422,16 +422,16 @@ added to the build using CMake's `add_subdirectory()`_ command, and the
-   ``-DZEPHYR_EXTRA_MODULES=/<path>/foo`` then the module given by the command
-   line variable :makevar:`ZEPHYR_EXTRA_MODULES` will take precedence.
+   ``-DEXTRA_ZEPHYR_MODULES=/<path>/foo`` then the module given by the command
+   line variable :makevar:`EXTRA_ZEPHYR_MODULES` will take precedence.
```
I think it's worth leaving a note that the old variable is still supported for backwards compat.

tejlmand replied:
I don't remember that we keep old variable names around in our docs when deprecating them, but i'm happy to do so in this case and for the `OVERLAY_CONFIG`.

mbolivar-nordic reviewed (changes requested):
What about the release notes?

tejlmand committed:
doc: release: 3.4: add note about EXTRA_<VAR> Zephyr build variables

Update release notes 3.4 with description of
EXTRA_CONF_FILE, EXTRA_DTC_OVERLAY_FILE, and EXTRA_ZEPHYR_MODULES.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> What about the release notes?

Updated.

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2507,76 +2538,61 @@ function(zephyr_get variable)
-    if(GET_VAR_MERGE)
-      list(APPEND ${variable} $ENV{${variable}}})
+    if(SYSBUILD)
+      get_property(sysbuild_name TARGET sysbuild_cache PROPERTY SYSBUILD_NAME)
+      get_property(sysbuild_main_app TARGET sysbuild_cache PROPERTY SYSBUILD_MAIN_APP)
```
This doesn't need to be ran multiple times in the for() loop

tejlmand replied:
correct, but I preferred keeping those lines close to the line below:
https://github.com/zephyrproject-rtos/zephyr/blob/53c2430dbe2a4c2ee0a2c536f2ae744ad7dbee5f/cmake/modules/extensions.cmake#L2555

if moving those two lines outside the for-loop then there would be a need one more `if(SYSBUILD) ... endif()` guard.
I consider the current approach more readable where the sysbuild code is located together.

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2507,76 +2538,61 @@ function(zephyr_get variable)
+        else()
+          set(${variable} ${scope_value} PARENT_SCOPE)
+
+          if("${scope}" STREQUAL "ENV" AND DEFINED current_${var}
+	     AND NOT "${current_${var}}" STREQUAL "$ENV{${var}}")
```
Adjust alignment here

nordicjm replied:
Something is not right here and I'm not sure if it's the file or github, in the file diff view, the "AND NOT" line is way left, in the comment view, it is right.

tejlmand replied:
fixed

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2507,76 +2538,61 @@ function(zephyr_get variable)
+  # This is needed to properly handle cases where we want to check value against
+  # environment value or when appending with the MERGE operation.
+  foreach(var ${GET_VAR_VAR})
+    set(current_${var} ${${var}})
```
I don't see the backup value being restored, only checked against the environmental value

tejlmand replied:
it's being restored through this loop:
https://github.com/zephyrproject-rtos/zephyr/blob/53c2430dbe2a4c2ee0a2c536f2ae744ad7dbee5f/cmake/modules/extensions.cmake#L2566-L2567
which goes through here:
https://github.com/zephyrproject-rtos/zephyr/blob/53c2430dbe2a4c2ee0a2c536f2ae744ad7dbee5f/cmake/modules/extensions.cmake#L2487-L2488

The warning itself is a special handling for the environment variable hiding a local one, as such a situation may indicate a user having something unexpected in environment.

nordicjm commented on cmake/modules/dts.cmake:
```diff
@@ -75,6 +75,12 @@ find_package(Dtc 1.4.6)
+# - EXTRA_DTC_OVERLAY_FILE: list of extra devicetree overlay files.
+#   This variable is is similar to DTC_OVERLAY_FILE but the files in
+#   EXTRA_DTC_OVERLAY_FILE will be applied after DTC_OVERLAY_FILE and
+#   thus files specified by EXTRA_DTC_OVERLAY_FILE has higher precedence.
```
*have

tejlmand replied:
fixed

nordicjm commented on cmake/modules/dts.cmake:
```diff
@@ -83,6 +89,9 @@ find_package(Dtc 1.4.6)
 # - DTS_SOURCE: the devicetree source file to use may be pre-set
 #   with this variable; otherwise, it defaults to
 #   ${BOARD_DIR}/${BOARD.dts}
+# - DTS_SOURCE: the devicetree source file to use may be pre-set
+#   with this variable; otherwise, it defaults to
+#   ${BOARD_DIR}/${BOARD.dts}
```
Duplicate comment, and above needs to be adjusted to `${BOARD}.dts`

tejlmand replied:
How did that get duplicated  😮 

And fixed old error 😉

nordicjm commented on doc/releases/release-notes-3.4.rst:
```diff
@@ -351,6 +351,17 @@ Build system and infrastructure
+  setting lookup.
+  ``EXTRA_CONF_FILE`` replaces ``OVERLAY_CONFIG``.
+  ``EXTRA_ZEPHYR_MODULES`` replaces ``ZEPHYR_EXTRA_MODULES``.
+  ``EXTRA_DTC_OVERLAY_FILE`` is new.
```
Add link to documentation page that describes it

tejlmand replied:
link added

stephanosio commented:
> > I count 147 uses of OVERLAY_CONFIG in `samples`, 37 in `tests` and 6 in `scripts` which will need updating
> 
> Correct, but those are within sample.yaml / testcase.yaml and used by `twister`. Because we keep support for the OVERLAY_CONFIG, then I prefer to wait with updating twister related files until we have this PR merged.
> 
> That way we ensure that true support for OVERLAY_CONFIG still works as expected, because twister still uses this, and then a followup PR adjusting sample.yaml / testcase.yaml to new variable name ensures that new name also works in all cases.
> 
> Finally it keeps number of files touched in this PR a bit smaller.
> 
> Note: some rst files are using OVERLAY_CONFIG, they should be updated in this PR, and I will do so.

FWIW, the use of `OVERLAY_CONFIG` in `extra_args` of `sample.yaml` and `testcase.yaml` is already deprecated:

https://github.com/zephyrproject-rtos/zephyr/blob/7d898edfc5c2e8dfa4bc158c5f74614f15dd833b/scripts/pylib/twister/twisterlib/config_parser.py#L229-L237

Instead, twister already has something similar to what is proposed here for the build system:

https://github.com/zephyrproject-rtos/zephyr/blob/7d898edfc5c2e8dfa4bc158c5f74614f15dd833b/scripts/pylib/twister/twisterlib/config_parser.py#L210-L214

https://github.com/zephyrproject-rtos/zephyr/blob/7d898edfc5c2e8dfa4bc158c5f74614f15dd833b/scripts/pylib/twister/twisterlib/config_parser.py#L223-L227

So, any explicit specifications of `OVERLAY_CONFIG` in `extra_args` should be updated to use these.

By the way, twister's `extra_conf_files` is actually for `CONF_FILE`, and instead it has `extra_overlay_confs` is for `OVERLAY_CONFIG`; so, after this PR is merged, we will either need to rename `extra_overlay_files` to `extra_extra_conf_files` or rethink how `extra_conf_files` works.

FYI @nashif @tristan-google @PerMac

tejlmand commented on doc/releases/release-notes-3.4.rst:
```diff
@@ -351,6 +351,18 @@ Build system and infrastructure
+* Enhanced ``zephyr_get`` CMake helper function to lookup multiple variables
+  and return the result in a variable of different name.
+
+* Introduce ``EXTRA_CONF_FILE``, ``EXTRA_DTC_OVERLAY_CONFIG``, and
```
> typo?

yep, thanks for spotting 👍
